core/vcpu.c

```diff
@@ -1504,8 +1464,14 @@ int vcpu_execute(struct vcpu_t *vcpu)
     if (htun->_exit_status == HAX_EXIT_IO) {
         handle_io_post(vcpu, htun);
     }
-    if (htun->_exit_status == HAX_EXIT_FAST_MMIO) {
-        handle_mmio_post(vcpu, (struct hax_fastmmio *)vcpu->io_buf);
+    // Continue until emulation finished
+    if (!em_ctxt->finished) {
+        switch (em_emulate_insn(em_ctxt)) {
+        case EM_EXIT_MMIO:
+            return HAX_EXIT_FAST_MMIO;
+        case EM_ERROR:
+            return HAX_EXIT;
```

Comment:
`HAX_EXIT` indicates success. In case of an error, probably you should call `hax_panic_vcpu`, set `err` to some error code, and `goto out`?